app/demo-loop/diagnostic/StepDiagnostic.i.hh

```diff
@@ -0,0 +1,159 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2021 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file StepDiagnostic.i.hh
+//---------------------------------------------------------------------------//
+
+#include <algorithm>
+#include "base/CollectionBuilder.hh"
+#include "physics/base/PhysicsTrackView.hh"
+#include "sim/SimTrackView.hh"
+
+namespace demo_loop
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from shared data.
+ */
+template<MemSpace M>
+StepDiagnostic<M>::StepDiagnostic(const ParamsDataRef& params,
+                                  SPConstParticles     particles,
+                                  size_type            num_tracks,
+                                  size_type            max_steps)
+    : params_(params), particles_(particles)
+{
+    CELER_EXPECT(params_);
+    CELER_EXPECT(particles_);
+    CELER_EXPECT(num_tracks > 0);
+    CELER_EXPECT(max_steps > 0);
+
+    StepDiagnosticData<Ownership::value, MemSpace::host> host_data;
+    host_data.max_steps     = max_steps;
+    host_data.num_particles = particles_->size();
+
+    // Initialize current number of steps in active tracks to zero
+    std::vector<size_type> zeros(num_tracks);
+    celeritas::make_builder(&host_data.steps)
+        .insert_back(zeros.begin(), zeros.end());
+
+    // Tracks binned by number of steps and particle type (indexed as
+    // particle_id * max_steps + num_steps)
+    zeros.resize(host_data.max_steps * host_data.num_particles);
+    celeritas::make_builder(&host_data.counts)
+        .insert_back(zeros.begin(), zeros.end());
+
+    data_ = host_data;
+    CELER_ENSURE(data_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the distribution of steps per track.
+ *
+ * This must be called after the interactors have been launched but before the
+ * post-processing.
+ */
+template<MemSpace M>
+void StepDiagnostic<M>::mid_step(const StateDataRef& states)
+{
+    StepDiagnosticDataRef<M> data_ref;
+    data_ref = data_;
+    count_steps(params_, states, data_ref);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get distribution of steps per track for each particle type.
+ *
+ * For i in [0, max_steps), steps[particle][i] is the number of tracks of the
+ * given particle type that took i + 1 steps.
+ */
+template<MemSpace M>
+std::unordered_map<std::string, std::vector<size_type>>
+StepDiagnostic<M>::steps()
+{
```

Comment:
This would be a good candidate for moving to a non-templated base class (if the Diagnostic class were a pure interface class). This sort of situation might be motivation for taking a "callback" style approach rather than an inheritance hierarchy (since it could allow greater code reuse).

Reply:
Good point, it's probably a good time to think about this now that we already have several diagnostics implemented.